File: src/app/data_processing.cpp
#include "app/data_processing.h"

#include <google/protobuf/util/json_util.h>
#include <tuple>

namespace Serializers
{
namespace App
{
Plugin::Plugin(const PbApp::Plugin& protobuf) {
    path = protobuf.path();
}

Plugin::operator PbApp::Plugin() const {
    PbApp::Plugin ret;
    ret.set_path(path);
    return ret;
}

std::string Plugin::serializeAsJsonString() const {
    PbApp::Plugin protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string Plugin::serializeAsProtobufString() const {
    PbApp::Plugin protobuf(*this);
    return protobuf.SerializeAsString();
}

bool Plugin::deserializeFromJsonString(const std::string& str) {
    PbApp::Plugin protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool Plugin::deserializeFromProtobufString(const std::string& str) {
    PbApp::Plugin protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

Signal::Signal(const PbApp::Signal& protobuf) {
    msg = protobuf.msg();
    fields = {protobuf.fields().begin(), protobuf.fields().end()};
}

Signal::operator PbApp::Signal() const {
    PbApp::Signal ret;
    ret.set_msg(msg);
    *(ret.mutable_fields()) = {fields.begin(), fields.end()};
    return ret;
}

std::string Signal::serializeAsJsonString() const {
    PbApp::Signal protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string Signal::serializeAsProtobufString() const {
    PbApp::Signal protobuf(*this);
    return protobuf.SerializeAsString();
}

bool Signal::deserializeFromJsonString(const std::string& str) {
    PbApp::Signal protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool Signal::deserializeFromProtobufString(const std::string& str) {
    PbApp::Signal protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

DataProcessing::DataProcessing(const PbApp::DataProcessing& protobuf) {
    plugins = {protobuf.plugins().begin(), protobuf.plugins().end()};
    resampledSignals = {protobuf.resampledsignals().begin(), protobuf.resampledsignals().end()};
}

DataProcessing::operator PbApp::DataProcessing() const {
    PbApp::DataProcessing ret;
    *(ret.mutable_plugins()) = {plugins.begin(), plugins.end()};
    *(ret.mutable_resampledsignals()) = {resampledSignals.begin(), resampledSignals.end()};
    return ret;
}

std::string DataProcessing::serializeAsJsonString() const {
    PbApp::DataProcessing protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string DataProcessing::serializeAsProtobufString() const {
    PbApp::DataProcessing protobuf(*this);
    return protobuf.SerializeAsString();
}

bool DataProcessing::deserializeFromJsonString(const std::string& str) {
    PbApp::DataProcessing protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool DataProcessing::deserializeFromProtobufString(const std::string& str) {
    PbApp::DataProcessing protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}
}
}

File: py/app/data_processing.py
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Dict

from app import data_processing_pb2
from google.protobuf.json_format import MessageToJson, Parse


@dataclass
class Plugin:
    path: str = ""
    
    _proto_message: data_processing_pb2.Plugin = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = data_processing_pb2.Plugin()

    def _populate_proto(self):
        self._proto_message.path = self.path

    @classmethod
    def _from_proto(cls, proto_message) -> "Plugin":
        return cls(
            path = proto_message.path,
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "Plugin":
        message = data_processing_pb2.Plugin()
        message.ParseFromString(data)
        return cls(
            path = message.path,
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "Plugin":
        message = data_processing_pb2.Plugin()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

@dataclass
class Signal:
    msg: str = ""
    fields: List[str] = field(default_factory=list)
    
    _proto_message: data_processing_pb2.Signal = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = data_processing_pb2.Signal()

    def _populate_proto(self):
        self._proto_message.msg = self.msg
        del self._proto_message.fields[:]
        for val in self.fields:
            self._proto_message.fields.append(val)

    @classmethod
    def _from_proto(cls, proto_message) -> "Signal":
        return cls(
            msg = proto_message.msg,
            fields=[str(val) for val in proto_message.fields],
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "Signal":
        message = data_processing_pb2.Signal()
        message.ParseFromString(data)
        return cls(
            msg = message.msg,
            fields = [str(val) for val in message.fields],
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "Signal":
        message = data_processing_pb2.Signal()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

@dataclass
class DataProcessing:
    plugins: List[Plugin] = field(default_factory=list)
    resampledSignals: List[Signal] = field(default_factory=list)
    
    _proto_message: data_processing_pb2.DataProcessing = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = data_processing_pb2.DataProcessing()

    def _populate_proto(self):
        del self._proto_message.plugins[:]
        for val in self.plugins:
            val._populate_proto()
            tmp = self._proto_message.plugins.add()
            tmp.CopyFrom(val._proto_message)
        del self._proto_message.resampledSignals[:]
        for val in self.resampledSignals:
            val._populate_proto()
            tmp = self._proto_message.resampledSignals.add()
            tmp.CopyFrom(val._proto_message)

    @classmethod
    def _from_proto(cls, proto_message) -> "DataProcessing":
        return cls(
            plugins=[Plugin._from_proto(val) for val in proto_message.plugins],
            resampledSignals=[Signal._from_proto(val) for val in proto_message.resampledSignals],
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "DataProcessing":
        message = data_processing_pb2.DataProcessing()
        message.ParseFromString(data)
        return cls(
            plugins = [Plugin._from_proto(val) for val in message.plugins],
            resampledSignals = [Signal._from_proto(val) for val in message.resampledSignals],
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "DataProcessing":
        message = data_processing_pb2.DataProcessing()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())
